### ISSUE TITLE
CheckpointInputPipelineHook not restoring the correct iterator state

### ISSUE THREAD

**Issue by Silb78dg** [tensorflow/tensorflow#24408]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOs / Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 1.11 / 1.12
- Python version: 3.5
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
1. CheckpointInputPipelineHook is not restoring the input checkpoint if there is multiple "estimator.train" method calls. Probably because the restore of the input checkpoint is done during the init of the hook. It restarts from the beginning of the dataset.
2. If both MirroredStrategy and CheckpointInputPipelineHook are used, one step is by passing when the input checkpoint is restored. It restarts from (state step + 1).

**Describe the expected behavior**
After restoring the input checkpoint, the train restart at the right dataset iterator step.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
In this sample code, I have one dataset (range between 0 and 7, batch of 1, 1 epoch), I train the model for 4 steps, stop it, restart for 4 steps the training by restoring the checkpoints (model and input).
Expected:
Train First Run: 4 steps -> features 0 to 3
Save checkpoint model-ckpt and input-ckpt 4
Restore model-ckpt and input-ckpt 4
Train Second Run: 4 steps -> features 4 to 7
Save checkpoint 8
`
import tensorflow as tf

sample_count = 8

def run(model_dir: str, reinit_hook: bool = False, distribution_strategy: bool = False, gpu_count: int = 1):
    # Config
    if distribution_strategy:
        session_config = tf.ConfigProto(allow_soft_placement=True)
        train_distribute = tf.contrib.distribute.MirroredStrategy(num_gpus=gpu_count)
    else:
        session_config = None
        train_distribute = None

    run_config = tf.estimator.RunConfig(
        model_dir=model_dir,
        session_config=session_config,
        train_distribute=train_distribute)

    # Estimator
    estimator = tf.estimator.Estimator(model_fn=model_fn, config=run_config)

    # Hook
    hooks = [tf.contrib.data.CheckpointInputPipelineHook(estimator)]

    # First Run 1/2
    estimator.train(input_fn=train_input_fn,
                    steps=sample_count // 2 // gpu_count,
                    hooks=hooks)

    # Reinit Hook
    if reinit_hook:
        hooks = [tf.contrib.data.CheckpointInputPipelineHook(estimator)]

    # Second Run 1/2
    estimator.train(input_fn=train_input_fn,
                    steps=sample_count // 2 // gpu_count,
                    hooks=hooks)

def train_input_fn():
    return tf.data.Dataset \
        .range(sample_count) \
        .map(lambda x: (x, x * 2)) \
        .batch(1) \
        .repeat(1)

def model_fn(features, labels, mode):
    input_layer = tf.cast(tf.reshape(features, [-1, 1]), tf.float32)
    expected_output = tf.cast(tf.reshape(labels, [-1, 1]), tf.float32)

    logit = tf.layers.dense(input_layer, 1, None, False)
    loss = tf.losses.mean_squared_error(expected_output, logit)

    logging_hook = tf.train.LoggingTensorHook(
        tensors=
        {
            "feature_value": features.name,
            "label_value": labels.name
        },
        every_n_iter=1)

    if mode == tf.estimator.ModeKeys.TRAIN:
        optimizer = tf.train.AdamOptimizer(0.001)
        train_op = optimizer.minimize(loss=loss, global_step=tf.train.get_global_step())
        return tf.estimator.EstimatorSpec(mode=mode,
                                          loss=loss,
                                          train_op=train_op,
                                          training_hooks=[logging_hook])

if __name__ == '__main__':
    tf.logging.set_verbosity(tf.logging.DEBUG)

    # Expected:
    # Train First Run: 4 steps -> features 0 to 3
    # Save checkpoint model-ckpt and input-ckpt 4
    # Restore model-ckpt and input-ckpt 4
    # Train Second Run: 4 steps -> features 4 to 7
    # Save checkpoint 8

    # run_1: 1 gpu / no distribution strategy / Only one init of the hook
    # Actual KO: dataset restarts from 0
    # 
    # Train First Run: 4 steps -> features 0 to 3
    # Save checkpoint model-ckpt and input-ckpt 4
    # Restore model-ckpt 4
    # Train Second Run: 4 steps -> features 0 to 3
    # Save checkpoint 8
    run(model_dir='output_run_1', reinit_hook=False, distribution_strategy=False)

    # run_2: 1 gpu / no distribution strategy / reinit of the hook between two trainings
    # Actual OK
    # Train First Run: 4 steps -> features 0 to 3
    # Save checkpoint model-ckpt and input-ckpt 4
    # Restore model-ckpt and input-ckpt 4
    # Train Second Run: 4 steps -> features 4 to 7
    # Save checkpoint 8
    run(model_dir='output_run_2', reinit_hook=True, distribution_strategy=False)

    # run_3:
    # Actual KO: dataset restarts from iterator state step + 1
    # Train First Run: 4 steps -> features 0 to 3
    # Save checkpoint model-ckpt and input-ckpt 4
    # Restore model-ckpt and input-ckpt 4
    # Train Second Run: 3 steps -> features 5 to 7
    # Save checkpoint 7
    run(model_dir='output_run_3', reinit_hook=True, distribution_strategy=True)

    # run_4:
    # Actual KO: dataset restarts from iterator state step + 1
    # Train First Run: 2 steps -> features 0 to 3
    # Save checkpoint model-ckpt and input-ckpt 2
    # Restore model-ckpt and input-ckpt 2
    # Train Second Run: 1 steps -> features 6 to 7
    # Save checkpoint 3
    run(model_dir='output_run_4', reinit_hook=True, distribution_strategy=True, gpu_count=2)
`

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by Silb78dg** [comment]
Hi all,
Any updates on this? Or any workaround?
I couldn't find a way to do preemption on distributed training.

Thanks.
Denis.

**Comment by guptapriya** [comment]
@rohan100jain is this something you can look into? (or someone else from tf.data?)

**Comment by rohan100jain** [comment]
Firstly, run_2 onwards (i.e. reinit_hook=True) is the way to go. 

So run_2 seems to work and in the non-distribution strategies case, things seem to be working fine.

For the distribution strategies case, the problem is that it uses MultiDeviceIterator (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/ops/multi_device_iterator_ops.py#L190) (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/multi_device_iterator_ops.cc#L43) under the hood who's state we currently don't checkpoint. This is something we need to fix on our end and will require a non-trivial amount of work. I currently don't have cycles to spend on this but contributions are welcome!

**Comment by Silb78dg** [comment]
@rohan100jain Thanks for the update. I will try to take a look. Is there any tests I can rely one?

**Comment by mohantym** [comment]
Hi @Silb78dg! 
Hi ! we are  checking to see if you are still looking for assistance in this issue.
Could you please try on latest stable version of  TF 2.6  and let us know if this is still an issue.Thanks!
